File: Python/Stack/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        lookup = {']':'[', ')':'(', '}':'{'}
        for c in s:
            if c == '(' or c == '[' or c == '{':
                stack.append(c)
            else:
                if len(stack)==0 or stack[-1]!=lookup[c]: return False
                stack.pop()
        return len(stack)==0

File: C++/Stack/20. Valid Parentheses.cpp
#include <stack>
#include <map>

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        map<char, char> lookup = {{']', '['}, {')', '('}, {'}', '{'}};
        for (char c : s) {
            if (c == '(' || c == '[' || c == '{') {
                stk.push(c);
            }
            else {
                if (stk.empty() || stk.top() != lookup[c]) return false;
                stk.pop();
            }
        }
        return stk.empty();
    }
};